pkg/bootstrap/docker/dockerhelper/client.go

```diff
@@ -0,0 +1,285 @@
+		return nil, ctxErr
+	}
+	return &info, err
+}
+
+func (c *dockerClient) ContainerList(options types.ContainerListOptions) ([]types.Container, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	containers, err := c.client.ContainerList(ctx, options)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return containers, err
+}
+
+func (c *dockerClient) ContainerCreate(config *container.Config, hostConfig *container.HostConfig, networkingConfig *network.NetworkingConfig, name string) (*types.ContainerCreateResponse, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerCreate(ctx, config, hostConfig, networkingConfig, name)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return &response, err
+}
+
+func (c *dockerClient) ContainerInspect(container string) (*types.ContainerJSON, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerInspect(ctx, container)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return nil, ctxErr
+	}
+	return &response, err
+}
+
+func (c *dockerClient) ContainerRemove(container string, options types.ContainerRemoveOptions) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	err := c.client.ContainerRemove(ctx, container, options)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return ctxErr
+	}
+	return err
+}
+
+func (c *dockerClient) ContainerLogs(container string, options types.ContainerLogsOptions, stdOut, stdErr io.Writer) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	response, err := c.client.ContainerLogs(ctx, container, options)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return ctxErr
+	}
+	if err != nil {
+		return err
+	}
+	defer response.Close()
+	return redirectResponseToOutputStream(stdOut, stdErr, response)
+}
+
+func (c *dockerClient) ContainerStart(container string) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	err := c.client.ContainerStart(ctx, container)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return ctxErr
+	}
+	return err
+}
+
+func (c *dockerClient) ContainerStop(container string, timeout int) error {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	err := c.client.ContainerStop(ctx, container, timeout)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return ctxErr
+	}
+	return err
+}
+
+func (c *dockerClient) ContainerWait(container string) (int, error) {
+	ctx, cancel := defaultContext()
+	defer cancel()
+	rc, err := c.client.ContainerWait(ctx, container)
+	if ctxErr := contextError(ctx); ctxErr != nil {
+		return 0, ctxErr
+	}
+	return rc, err
+}
+
+func (c *dockerClient) CopyToContainer(container string, dest string, src io.Reader, options types.CopyToContainerOptions) error {
+	return c.client.CopyToContainer(context.Background(), container, dest, src, options)
+}
+
+func (c *dockerClient) CopyFromContainer(container string, src string, dest io.Writer) error {
```

Comment:
Suggest you return all the values from c.client.CopyFromContainer straight.  Then you get rid of the io.Copy here, and the goroutine, pipe and io.Copy in newContainerDownloader.